perfkitbenchmarker/providers/openstack/os_virtual_machine.py

```diff
@@ -33,214 +43,328 @@
+  uploaded_keypair_set = set()
+  deleted_keypair_set = set()
+  created_server_group_dict = {}
+  deleted_server_group_set = set()
+
+  def __init__(self, vm_spec):
+    """Initialize an OpenStack virtual machine.
+
+    Args:
+      vm_spec: virtual_machine.BaseVirtualMachineSpec object of the vm.
+    """
+    super(OpenStackVirtualMachine, self).__init__(vm_spec)
+    self.key_name = 'perfkit_key_%s' % FLAGS.run_uri
+    self.user_name = FLAGS.openstack_image_username
+    self.image = self.image or self.DEFAULT_IMAGE
+    # FIXME(meteorfox): Remove --openstack_public_network and
+    # --openstack_private_network once depreciation time has expired
+    self.network_name = (FLAGS.openstack_network or
+                         FLAGS.openstack_private_network)
+    self.floating_ip_pool_name = (FLAGS.openstack_floating_ip_pool or
+                                  FLAGS.openstack_public_network)
+    self.id = None
+    self.boot_volume_id = None
+    self.server_group_id = None
+    self.floating_ip = None
+    self.firewall = None
+    self.public_network = None
+
+  @property
+  def group_id(self):
+    """Returns the security group ID of this VM."""
+    return 'perfkit_sc_group'
+
+  def _CreateDependencies(self):
+    """Validate and Create dependencies prior creating the VM."""
+    self._CheckPrerequisites()
+    self.firewall = os_network.OpenStackFirewall.GetFirewall()
+    self.public_network = os_network.OpenStackFloatingIPPool(
+        self.floating_ip_pool_name)
+    self._UploadSSHPublicKey()
+    self.firewall.AllowICMP(self)  # Allowing ICMP traffic (i.e. ping)
+    self.AllowRemoteAccessPorts()
+
+  def _Create(self):
+    """Creates an OpenStack VM instance and waits until it is ACTIVE."""
+    if FLAGS.openstack_boot_from_volume:
+      vol_name = '%s_volume' % self.name
+      disk_resp = os_disk.CreateVolume(self, vol_name, self.image)
+      self.boot_volume_id = disk_resp['id']
+      os_disk.WaitForVolumeCreation(self, self.boot_volume_id)
+    self._CreateInstance()
+
+  @vm_util.Retry(max_retries=4, poll_interval=2)
+  def _PostCreate(self):
+    self._SetIPAddresses()
+
+  def _Delete(self):
+    if self.id is None:
+      return
+    self._DeleteInstance()
+    if self.floating_ip:
+      self.public_network.release(self, self.floating_ip)
+    if self.server_group_id:
+      self._DeleteServerGroup()
+    if self.boot_volume_id:
+      os_disk.DeleteVolume(self, self.boot_volume_id)
+      self.boot_volume_id = None
+
+  def _DeleteDependencies(self):
+    """Delete dependencies that were needed for the VM after the VM has been
+    deleted."""
+    self._DeleteSSHPublicKey()
+
+  def _Exists(self):
+    if self.id is None:
+      return False
+
+    show_cmd = os_utils.OpenStackCLICommand(self, 'server', 'show', self.id)
+    stdout, _, _ = show_cmd.Issue(suppress_warning=True)
+    try:
+      resp = json.loads(stdout)
+    except ValueError:
+      return False
+
+    return resp is not None
```

Comment:
Can `resp` ever be `None` here? I don't understand under what circumstances `json.loads` would return `None`. If I don't understand `json.loads`, please correct me, but otherwise I think you can just return from inside the `try` block.